### PR TITLE
[feat] Implement `DisjunctionMaxQuery` and refactor `ScoreCombiner`

### PR DESCRIPTION
**Pull request by ppodolsky**
Here I reproduced ES [DisMax Query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-dis-max-query.html)

Not sure if my approach is well aligned with general design, so any suggestions to redesign it are welcomed.
I've decided to reuse `BooleanQuery` and built `DisjuctionMaxQuery` on it. For the purpose I had refactored `ScoreCombiner` part. Now `ScoreCombiner` creation is done through passed lambda but not through static dispatching. It allowed to configure `ScoreCombiner` properties. In particular, `DisjunctionMaxCombiner`s with various `tie_breaker` parameters are possible now.

### CODE REVIEW

**Comment by ppodolsky** [comment]
@PSeitz Are there any interest in such kind of queries?

**Comment by PSeitz** [comment]
@ppodolsky sorry I missed that review. Yes support for such queries is definitely interesting.

Could you check the performance impact on `bench_union_3_low` for removing the static dispatch?

**Comment by ppodolsky** [comment]
@PSeitz Checked, there is no any confident difference between two branches (cut few lines below to make output clearer):
```
pasha@home:~/tantivy$ git checkout upstream/main
HEAD is now at 8e773ade7 Merge pull request #1444 from quickwit-oss/add-async-doc-freq
pasha@home:~/tantivy$ cargo +nightly bench -F=unstable --color=always --package tantivy --lib query::union::bench::bench_union_3_high -- --exact
   Compiling tantivy v0.18.0 (/home/pasha/tantivy)
    Finished bench [optimized] target(s) in 57.24s
     Running unittests src/lib.rs (target/release/deps/tantivy-80ba2a9ba3804a7a)
running 1 test
test query::union::bench::bench_union_3_high                                                                             ... bench:     281,312 ns/iter (+/- 9,280)
pasha@home:~/tantivy$ git checkout origin/feature/dismax
Previous HEAD position was 8e773ade7 Merge pull request #1444 from quickwit-oss/add-async-doc-freq
HEAD is now at 71041b231 [fix] Fix bench
pasha@home:~/tantivy$ cargo +nightly bench -F=unstable --color=always --package tantivy --lib query::union::bench::bench_union_3_high -- --exact
   Compiling tantivy v0.18.0 (/home/pasha/tantivy)
    Finished bench [optimized] target(s) in 50.65s
     Running unittests src/lib.rs (target/release/deps/tantivy-21841e494d11c590)
running 1 test
test query::union::bench::bench_union_3_high                                                                             ... bench:     280,469 ns/iter (+/- 22,894)
```

**Comment by ppodolsky** [comment]
@PSeitz Is there anything else I can do in this PR?

File: src/query/union.rs [hunk 1]
```diff
@@ -90,6 +62,31 @@ fn refill<TScorer: Scorer, TScoreCombiner: ScoreCombiner>(
 }
 
 impl<TScorer: Scorer, TScoreCombiner: ScoreCombiner> Union<TScorer, TScoreCombiner> {
+    pub(crate) fn build(
+        docsets: Vec<TScorer>,
+        score_combiner_fn: impl Fn() -> TScoreCombiner,
```
Comment (PSeitz) [inline src/query/union.rs]: This can be a `FnOnce`

**Review by PSeitz — APPROVED** [review]
LGTM, except minor comment about FnOnce

File: src/query/union.rs (hunk 1 above)
Comment (ppodolsky) [reply src/query/union.rs]: @PSeitz Done

**Pull request merged**